### PR TITLE
upgrade deepmerge, add all method, add tests

### PR DESCRIPTION
**Pull request by bernardmcmanus**
* Upgrade deepmerge to support all object types
* Add all method to merge an array of config objects
* Add tests and coverage check
* Bump major version

### CODE REVIEW

File: .babelrc.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
```
Comment (eliperelman) [inline .babelrc.js:1]: Can you change the indentation of this file to 2 spaces?

**Review by eliperelman — CHANGES_REQUESTED** [review]
I can't even begin to describe how awesome this PR is! Thank you so much for doing this!

I just have one small request for spacing on the babelrc and I think we can merge.

File: .babelrc.js (hunk 1 above)
Comment (bernardmcmanus) [reply .babelrc.js:1]: 👍 sure thing

**Comment by bernardmcmanus** [comment]
@eliperelman so I was just thinking - how do you feel about changing the merge logic slightly so it aligns with babel's behavior? Babel merges plugins and presets but leaves options intact, which I think makes sense in most cases.

<details>
<summary><b>.babelrc.js</b></summary>

```js
module.exports = {
  presets: [
    [
      '@babel/env',
      {
        targets: {
          node: 'current'
        }
      }
    ]
  ],
  plugins: [
    '@babel/plugin-proposal-object-rest-spread',
    ['module:fast-async', {
      runtimePattern: null,
      useRuntimeModule: false
    }]
  ]
};
```
</details>

<details>
<summary><b>babel-test.js</b></summary>

```js
const Path = require('path');
const babel = require('@babel/core');

const babelMerge = require('./src');

function relativeToNodeModules(path) {
  return Path.relative(Path.resolve('node_modules'), path);
}

function simplifyBabelConfig({ file, options }) {
  const key = relativeToNodeModules(file.resolved);
  return options ? [key, options] : key;
}

function simplifyPlainConfig(value) {
  return typeof value === 'string'
    ? relativeToNodeModules(value)
    : [relativeToNodeModules(value[0]), value[1]]
}

const overrides = {
  presets: [
    [
      '@babel/env',
      {
        targets: {
          browsers: ['>= 0.25%', 'not dead']
        }
      }
    ]
  ],
  plugins: [
    ['module:fast-async', { spec: true }],
    '@babel/proposal-class-properties',
    ['@babel/proposal-object-rest-spread', {
      useBuiltIns: true
    }]
  ]
};

let presets;
let plugins;

({ options: { presets, plugins } } = babel.loadPartialConfig(overrides));

console.log('\nbabel');
console.dir(
  {
    presets: presets.map(simplifyBabelConfig),
    plugins: plugins.map(simplifyBabelConfig)
  },
  { colors: true, depth: null }
);

({ presets, plugins } = babelMerge(require('./.babelrc'), overrides));

console.log('\nbabel-merge');
console.dir(
  {
    presets: presets.map(simplifyPlainConfig),
    plugins: plugins.map(simplifyPlainConfig)
  },
  { colors: true, depth: null }
);
```
</details>

<br />

```
> node babel-test.js

babel
{ presets:
   [ [ '@babel/preset-env/lib/index.js',
       { targets: { browsers: [ '>= 0.25%', 'not dead' ] } } ] ],
  plugins:
   [ [ 'fast-async/plugin.js', { spec: true } ],
     '@babel/plugin-proposal-class-properties/lib/index.js',
     [ '@babel/plugin-proposal-object-rest-spread/lib/index.js',
       { useBuiltIns: true } ] ] }

babel-merge
{ presets:
   [ [ '@babel/preset-env/lib/index.js',
       { targets: { node: 'current', browsers: [ '>= 0.25%', 'not dead' ] } } ] ],
  plugins:
   [ [ '@babel/plugin-proposal-object-rest-spread/lib/index.js',
       { useBuiltIns: true } ],
     [ 'fast-async/plugin.js',
       { runtimePattern: null, useRuntimeModule: false, spec: true } ],
     '@babel/plugin-proposal-class-properties/lib/index.js' ] }
```

**Comment by eliperelman** [comment]
@bernardmcmanus so is the intent that any options passed to the plugin or preset would be completely replaced instead of merged?

**Comment by bernardmcmanus** [comment]
@eliperelman yes - that's how babel handles config merging and I think it makes sense in most cases.

Or it might be nice to make that behavior configurable:

```js
merge(ours, theirs, {
  strategy: 'theirs', // or 'ours', 'combine', etc.
  // ...
});
```

**Comment by eliperelman** [comment]
@bernardmcmanus I think what you said originally is fine, let's just do a replace for now since that is synonymous with Babel. Is that what your current implementation here has? If so, I'll review again.

**Comment by bernardmcmanus** [comment]
@eliperelman Sounds good. It does not, I have those changes locally but need to update tests / docs etc. I should have some time later this afternoon.

**Review by eliperelman — APPROVED** [review]
INCREDIBLE! 💯 🎉

**Comment by eliperelman** [comment]
Published in v2.0.0.

**Comment by bernardmcmanus** [comment]
@eliperelman I totally forgot to update the README with new behavior 😬 I'll submit a new PR

**Comment by eliperelman** [comment]
@bernardmcmanus no worries, thanks!

**Pull request merged**